core/jobs/types/model_property.py

```diff
@@ -31,7 +38,14 @@ class ModelProperty:
 
     __slots__ = ('_model_kind', '_property_name')
 
-    def __init__(self, model_class, property_obj):
+    def __init__(
+        self,
+        model_class: Type[base_models.BaseModel],
+        property_obj: Union[
+            datastore_services.Property,
+            Callable[[base_models.BaseModel], str]
+        ]
```

Comment:
Why is this a union?

Reply:
The `propert_obj` argument can accept two types of property:
1. A property that is of type `datastore_services.Property`.
2. A python property.

However, if we look at the code of the constructor we can find that `property_obj` can only accept `id` python property. https://github.com/oppia/oppia/blob/ed2f6993a118527e169dccc4b44a39b81bebdf63/core/jobs/types/model_property.py#L55
and all types of property which are derived from `datastore_services.Property`. https://github.com/oppia/oppia/blob/ed2f6993a118527e169dccc4b44a39b81bebdf63/core/jobs/types/model_property.py#L58
So, The type of `id` python property is `Callable[[base_models.BaseModel], str]`, because return value of id is str and this property is decorated using a class ( pyton's property class ). And the type of  datastore_services.Property's property is `datastore_services.Property`. That's why union is used.

Reply:
Done!!